Mathlib/Analysis/Convex/Radon.lean

```diff
@@ -0,0 +1,232 @@
+        simp; constructor; exact h1; exact h_neg
+      simp at h4_aux; simp; linarith
+    constructor
+    linarith
+    exact h_J_neg
+
+    have h_I_pos : weights_sum_I > 0 := by
+      have h4_aux : weights_sum_I ≥ w nonzero_w_index := by
+        apply Finset.single_le_sum
+        intros i hi
+        rw [Finset.mem_filter] at hi; exact hi.right
+        simp; constructor; exact h1; exact LT.lt.le h_pos
+      linarith
+    constructor
+    exact h_I_pos
+    linarith
+  rcases h4 with ⟨h_I_pos, h_J_neg⟩
+
+  let w' : ι → 𝕜 := fun i => if w i ≥ 0
+    then (w i) / weights_sum_I
+    else -(w i) / weights_sum_I
+  let p : E := ∑ i in I, (w' i) • (f i) -- point of intersection
+
+  have h5_I : ∑ i in I, w' i = 1 := by
+    let w'' : ι → 𝕜 := fun i => (w i) / weights_sum_I
+    have h5_aux : ∑ i in I, w' i = ∑ i in I, w'' i := by
+      apply Finset.sum_congr; rfl
+      intros i hi
+      rw [Finset.mem_filter] at hi
+      rcases hi with ⟨_, hh⟩
+      simp
+      contrapose; intros
+      exact Iff.mpr not_lt hh
+    rw [h5_aux]
+    simp
+    rw [←Finset.sum_div]
+    apply div_self
+    exact ne_of_gt h_I_pos
+
+  have h5_J : ∑ i in J, w' i = 1 := by
+    let w'' : ι → 𝕜 := fun i => (w i) / weights_sum_J
+    have h5_aux1 : weights_sum_I = -weights_sum_J := by
+      exact Iff.mp add_eq_zero_iff_eq_neg h3
+    have h5_aux2 : ∑ i in J, w' i = ∑ i in J, w'' i := by
+      apply Finset.sum_congr; rfl
+      intros i hi
+      rw [Finset.mem_filter] at hi
+      rcases hi with ⟨_, h_neg⟩
+      simp
+      split_ifs
+      exfalso; linarith
+      simp_rw [h5_aux1, div_neg]
+      ring
+    rw [h5_aux2]; simp
+    rw [←Finset.sum_div]
+    apply div_self
+    exact ne_of_lt h_J_neg
+
+  have h6 : ∀ v, w' v ≥ 0 := by
+    intro; simp; split_ifs
+    repeat {apply div_nonneg; linarith; linarith}
+
+  let p' : E := (Finset.sum J fun i => (w' i) • (f i))
+  have h7 : p = p' := by
+    suffices h7_suf1 : p -ᵥ p' = 0 by
+      exact eq_of_vsub_eq_zero h7_suf1
+
+    rw [Finset.weightedVSub_eq_weightedVSubOfPoint_of_sum_eq_zero s1 w f h_wsum 0]
+      at h_vsum
+    simp at h_vsum; simp
+
+    let g : ι → E := fun i =>  weights_sum_I⁻¹ • ((w i) • (f i))
+    have h7_aux2 : ∑ i in I, (w' i) • (f i) = ∑ i in I, g i := by
+      apply Finset.sum_congr; rfl
+      intros i hi
+      rw [Finset.mem_filter] at hi
+      rcases hi
+      simp
+      ring_nf; split_ifs
+      rw [mul_comm, mul_smul]
+      exfalso; linarith
+
+    have h7_aux3 : ∑ i in J, (w' i) • (f i) = -∑ i in J, g i := by
+      rw [←Finset.sum_neg_distrib]
+      apply Finset.sum_congr; rfl
+      intros i hi
+      rw [Finset.mem_filter] at hi
+      rcases hi
+      simp
+      ring_nf; split_ifs
+      exfalso; linarith
+      rw [mul_comm, ←neg_smul, ←neg_mul, mul_smul]
+
+    rw [h7_aux2, h7_aux3]
+    simp
+    have h7_aux4 := Finset.sum_filter_add_sum_filter_not s1 (fun i => 0 ≤ w i) g
+    simp at h7_aux4; rw [h7_aux4]
+    rw [←Finset.smul_sum, h_vsum, smul_zero]
```

Comment:
- As mentioned above, you can completely avoid `VSub` here.
- Again, no need for an auxiliary function here
- The tactic `congr : 1` in my code generates two goals. The `<;>` combinator after it tells it that the next tactic should be run in both cases.
```suggestion
    apply eq_of_sub_eq_zero
    simp_rw [div_eq_inv_mul, mul_smul, ← smul_sum, ← smul_sub, sub_eq_add_neg, ← sum_neg_distrib]
    apply smul_eq_zero_of_right
    rw [← h_vsum, ← Finset.sum_filter_add_sum_filter_not s1 (fun i => 0 ≤ w i) (fun i ↦ w i • f i)]
    congr 1 <;>
    refine sum_congr (by simp only [not_le]) fun i hi ↦ ?_
    · rw [mem_filter] at hi
      rw [abs_eq_self.mpr hi.2]
    · rw [mem_filter, not_le] at hi
      rw [abs_eq_neg_self.mpr hi.2.le, neg_smul, neg_neg]
```